test-suite/Math/NumberTheory/Moduli/DiscreteLogarithmTests.hs

```diff
@@ -0,0 +1,52 @@
+{-# LANGUAGE ScopedTypeVariables #-}
+{-# OPTIONS_GHC -fno-warn-type-defaults #-}
+module Math.NumberTheory.Moduli.DiscreteLogarithmTests
+  ( testSuite
+  ) where
+
+import Data.Maybe
+import Numeric.Natural
+import Test.Tasty
+import Data.Semigroup
+import Data.Proxy
+import GHC.TypeNats.Compat
+
+import Math.NumberTheory.Moduli.Class
+import Math.NumberTheory.Moduli.PrimitiveRoot
+import Math.NumberTheory.Moduli.DiscreteLogarithm
+import Math.NumberTheory.ArithmeticFunctions (totient)
+import Math.NumberTheory.TestUtils
+
+-- | Ensure 'discreteLogarithm' returns in the appropriate range.
+discreteLogRange :: Positive Natural -> Integer -> Integer -> Bool
+discreteLogRange (Positive m) a b =
+  case someNatVal m of
+    SomeNat (_ :: Proxy m) -> fromMaybe True $ do
+      a' <- isPrimitiveRoot (fromInteger a :: Mod m)
+      b' <- isMultElement (fromInteger b)
+      return $ discreteLogarithm a' b' < totient m
```

Comment:
It is worth to check that it is also `>= 0`.

Reply:
The return type of `discreteLogarithm` is `Natural`, so it cannot be negative.

Reply:
OK, then it is fine.